### DISCUSSION TITLE
Should we mention backing up the master xpub?

### DISCUSSION THREAD

**Discussion by BenWestgate** [BlockstreamResearch/codex32#65] [Ideas]
Reading [last week's blog post](https://blog.blockstream.com/codex32-a-shamir-secret-sharing-scheme/):

> **Verify the coins are still there**. There is not much an offline backup can do to check the state of the blockchain. Instead, use an online watch-only wallet to monitor the blockchain for any movement of your coins, without ever needing access to secret data.

This only works without accessing secret data if we don't lose the master extended public key. It'd be a shame to need to expose secrets to electronics again just to watch the wallet. Perhaps we should mention somewhere the master extended public key may be backed up by codex32 as well for very long-term or high-value storage? And maybe have a standard human-readable part for this application.

maybe `mxpub` or `xpub` unfortunately has to be 513 bit data because we backup bip32 seeds instead of master private key data so chaincode has to be stored. I'd probably do this for a 50 year wallet.

**Comment by apoelstra** [comment]
The master xpub should be stored inside a descriptor (which gives all the necessary information to scan for coins), and descriptors already have a BCH code checksum on them. So we don't need to use codex32 for this.

But I am certainly open to adding text somewhere recommending that people store their descriptors somewhere, and explain that there is a privacy but not a key-loss risk, so they should probably be more liberal in making copies than they would for their seed data.

**Reply by BenWestgate** [reply]
Agree the descriptor should be stored if not using standard derivation paths. 

But the privacy risk is why I suggested codex32 for the actual extended key itself. Then it can be split at threshold 2 or encrypted by an unused share. That would correct more errors than the descriptor checksum too. (And be hand verifiable? But I think descriptor checksums can be hand computed)

**Reply by apoelstra** [reply]
Ah, I see!

My feeling is that codex32 is really not designed for this (the short checksum has a length of 93; the long one has length 1023. And even 1023 may not be long enough for a descriptor. The descriptor alphabet is different from the bech32 alphabet (descriptors have a 96-char alphabet which is broken down into the 32-element bech32 alphabet internally) which would complicate all the SSS processes.

And in the end I feel that this is something of a niche use case.

I'm not, in principle, opposed to this, but it would be a ton of work (and I don't want to do the work :)).

**Reply by BenWestgate** [reply]
Don't use codex32 for the scripts and paths.

Something like this (with the checksum) is correct:
![image](https://github.com/BlockstreamResearch/codex32/assets/73506583/c002165a-ce9a-4571-8241-43d9f9e3d19a)

"@0" and "@1" are master extended public keys. If someone backs this up on paper, to watch without putting private keys on electronics, some privacy would be nice to not reveal the whole wallet balance from a single compromised location.

Here's an example:

Master Seed shares:
Share with index `d`: `ms13cashd0wsedstcdcts64cd7wvy4m90lm28w4ffupqs7rm`
Share with index `e`: `ms13casheekgpemxzshcrmqhaydlp6yhms3ws7320xyxsar9`
Share with index `f`: `ms13cashf8jh6sdrkpyrsp5ut94pj8ktehhw2hfvyrj48704`


master node xprv: `xprv9s21ZrQH143K266qUcrDyYJrSG7KA3A7sE5UHndYRkFzsPQ6xwUhEGK1rNuyyA57Vkc1Ma6a8boVqcKqGNximmAe9L65WsYNcNitKRPnABd`

master node xpub:
`xpub661MyMwAqRbcEaBJaePELgFazHwoZVsyET156B39z5nykBjFWUnwn4dVhgSFRArDX4hWnnnG1gwGEeMju7qjh6YRQ3DwpLn2np518856Ry5`

Master xpub hex: `0488b21e00000000000000000003054649d15eef11b33570bad93a9520e0ef4039f4b9834a6d297f4a76cb2c5803a504f4cc8ab801639828237d269c072a3346f1470387d7da9d58acacc9f74922`
Master chaincode: `0x03054649d15eef11b33570bad93a9520e0ef4039f4b9834a6d297f4a76cb2c58`
Master pubkey data: `0x03a504f4cc8ab801639828237d269c072a3346f1470387d7da9d58acacc9f74922`

xpub @0 data (chaincode || pubkey data) to encode in codex32:
`0x03054649d15eef11b33570bad93a9520e0ef4039f4b9834a6d297f4a76cb2c5803a504f4cc8ab801639828237d269c072a3346f1470387d7da9d58acacc9f74922`


Secret extended public key @0 in codex32 format:
`xpub@012cashsqvz5vjw3tmh3rve4wzadjw54yrsw7spe7jucxjnd99l55akt93vq8fgy7nxg4wqpvwvzsgmay6wqw23ngmc5wqu86ldf6k9v4nylwjfznguwxye9fqxagda`

xpub Share with index `c`: `xpub@012cashcdy4xpurmmhd42uqpa3ktl93m0zvqxh5e582r7gmfx9fdugcaq6rz2henhazmfzd3chshcq8xzcpdruga2925prepz02qyg5p42mrrglqxy6vfsl0r8q0kvv`
xpub Share with index `d`: `xpub@012cashd4352ecmpcu5tvl9uzwl68vulx3yys46ewfcfg3wh69mgfxu0dn5djsfj3euypf4j9jced40vnhk4mwde9zx5kl5s5vgxr0ye400tm3d0g0tr7dw4e93r3lu`
xpub Share with index `e`: `xpub@012cashea9vw3rnsezypw7xhsuc40tqyp8ml9v8el5k0ap6af94tam8p35eg6272nzkk7hanw38yeps2ujmanqwzqcz5mvglpd3ye4n34vr5npr243yxnxpygu78tft`
xpub Share with index `f`: `xpub@012cashf84trtlfys6efuhaklnwe4a2ch92ruzyethnsyu2409ssdcmymz8vql4dp77e008602k8f3p4sk98f947yd0592zncylkvqct4h59fuxwkfgzd8dsujj27tq`

The cleartext descriptor `sh(sortedmulti(@0/**,@1/**))` stored with each xpub share now contains much less privacy sensitive data and can be recovered from paper without putting private keys on electronics. As an example, the threshold is set lower and the copies higher than the master seed itself, reflecting its lower confidentiality and greater availability needs. I reused identifier and indices to help associate the seed & xpub pair.

These `xpub@0` strings are a real working example I made with my python library.

**Reply by apoelstra** [reply]
Ok, sure, if you split only the xpub itself then you avoid the limitations of the checksum.

But this still feels overly complicated to me, and a bad tradeoff between the value and the complexity of the complete backup scheme. I'd be fine saying something like "you can do this if you want" or even casually recommending it, but I don't like the idea of doing extra work to support it.

**Reply by BenWestgate** [reply]
It'd be simpler (for users, not us) and give more value to encrypt or secret split the entire descriptor.

Maybe I'll do this work someday. I've avoided multi-sig in the past because of its pubkey privacy single point of failure.

If you weren't comfortable storing your descriptor backups in plaintext, what would you do today?

gpg encrypt them to yourself and cosigners? use a shamir's secret sharing tool like [libgfshare](https://packages.debian.org/trixie/libgfshare-bin)? use par2 to make parity files small enough 2 are needed to restore the descriptor?
If gpg, then secret split the decryption key so descriptor.txt.gpg can be saved and printed everywhere?

**Reply by apoelstra** [reply]
>If you weren't comfortable storing your descriptor backups in plaintext, what would you do today?

Well, the descriptor itself (without xpub data) I think is fine to be public. (In fact you can find the descriptor that I'm using by reading the source code of [ICBOC](https://github.com/apoelstra/icboc) which I used to set up my current cold wallet.) If I *really* didn't want it to be written down though I'd probably gpg-encrypt it to myself and multiple other people whose contact information I could put into my estate-planning docs.

For the xpub itself, if I didn't want to write it down then I'd just not. It can be rederived from the secret data if need be.

But TBH I find this scenario a little hard to imagine because it makes it so much harder to recover my coins (even to determine whether one is on the right track) in the case that I'm dead or incapacitated. So to make the privacy tradeoff worthwhile I'd need to be in a situation where I was like, actively worried about an organized and targeted effort to identify my coins, by somebody who knew me well enough (or had enough state-like powers) to dig through my personal documents.

**Reply by BenWestgate** [reply]
> For the xpub itself, if I didn't want to write it down then I'd just not. It can be rederived from the secret data if need be.

By avoidable electronics. Or are you saying to have so many copies of the descriptor that it's more likely to lose the master seed than the descriptor?

Say the IRS raids your home, finds the descriptor and claims there is unreported income or transactions to ISIS/Russia on it. Or lets say you used it to accept dark web payments for drug dealing. This is not safe to be clear text.

I'm also imagining a situation where Bitcoin is "illegal" to use, in which case a descriptor with recent transactions would be incriminating evidence.

Does gpg encrypting it to yourself really protect you? They would jail you until you give the passphrase to your private key. 
What about encrypting it so any two recipients private keys were needed to decrypt? Trying to avoid the privacy single point of failure since for some it is a matter of life and death.

**Reply by apoelstra** [reply]
>By avoidable electronics

Yes :).

>Say the IRS raids your home, finds the descriptor and claims there is unreported income or transactions to ISIS/Russia on it.

Well, using hardened derivations will prevent this case. As well as the case where "recent" transactions are detected, or any transactions.

I guess you would like a fully-derived descriptor, which could be used for watch-only purposes, but which *also* does not require electronic computers to obtain from secret data, and which *also* is stored using SSS. This does sound plausible, and I'd be happy to provide some pointers or basic support for such people, but I think for the vast majority of users (including myself) it's likely to do more harm than good.

>Does gpg encrypting it to yourself really protect you? They would jail you until you give the passphrase to your private key.

They can't get an open warrant for all gpg-encrypted data. There would have to be some clear indication that the encrypted data was an actual descriptor. (Or maybe we are in a mad max world with no warrants or due process.)

**Reply by BenWestgate** [reply]
> Well, using hardened derivations will prevent this case. As well as the case where "recent" transactions are detected, or any transactions.

By this you mean redacting your own child extended public key from a multisig descriptor? I think this depends on script type, OP_CHECKMULTISIG puts signer public keys on blockchain, linking recent spends (but not deposits) if scanning the blockchain. The new taproot multi_a type would make this perfectly private.

That sounds easy enough to do and hard enough to screw up, especially when codex32 IDs match the descriptor's key origin information fingerprint for the redacted xpub. (This ends up being useful so often, for so many things (even the CodexQR benefits), it should be the Recommended ID if no strong reason exists to change it.)

It **also** forms a sort of threshold 2 "secret sharing" between two co-signers (who both redacted their own child xpub but not others) where combining two has enough information to watch the wallet without putting private keys on electronics again.

This is quite slick, **would you recommend it?** Private keys touch electronics to restore descriptor by yourself, two co-signers required to restore descriptor without private keys on electronics. Only works with multi_a() scripts.

Only draw back seems to be co-signers can't store the descriptor in a public location and get strong privacy.

The way I'd get around that is also encrypt the descriptor to each co-signer. [BIP85](https://github.com/bitcoin/bips/blob/master/bip-0085.mediawiki#rsa) has a standard for recovering RSA keys from a BIP32 root key. That creates an encrypted descriptor that can be stored anywhere and redacted copies that need to be stored with some security and in different locations from your co-signers' redacted copies. If people keep a passphrase encrypted backup of their RSA private key, they can watch restore without _bitcoin_ private keys on electronics. If the RSA key is lost, it can be restored from the codex32 secret. You might deliberately lose it in a mad max scenario to make watching require compromising multiple locations and/or people not just a passphrase.

**Reply by apoelstra** [reply]
>This is quite slick, would you recommend it? Private keys touch electronics to restore descriptor by yourself, two co-signers required to restore descriptor without private keys on electronics. Only works with multi_a() scripts.

Yes, I think this is fine, because it's a "public 2-of-2" but "private 1-of-2", in the sense that either party **could** get the full descriptor back, as long as they were willing to use their secret data.

It doesn't generalize well if you're explicitly trying to use it as a sort of secret sharing scheme. But I *do* think it's reasonable to recommend "to get a private descriptor, redact your own keys". The fact that then any two parties will be able to reconstruct the xpubs feels like a curiosity to me rather than a deliberate part of the scheme....

> Only draw back seems to be co-signers can't store the descriptor in a public location and get strong privacy.

...hmm, yeah...this actually could be a bit surprising. Two co-signers might upload their descriptors to different public places, without realizing that the other one did, and then they lose their privacy. So maybe this is too footgunny.

I use this strategy for my own descriptors, but in my case I'm the only party so I am not vulnerable to this.

Thanks for the tip about BIP 85 BTW! I have wanted to do this but didn't realize there was a standard way.

> The way I'd get around that is also encrypt the descriptor to each co-signer. BIP85 has a standard for recovering RSA keys from...

Ok **in conclusion** I think this is a much better idea than suggesting people redact their keys. They should instead store their full watch-only descriptor, but gpg-encrypt it to their own key using BIP85. For the avoidance of doubt, we should specify that they use the *first key* that they own and they should use a specific BIP32 path from it.

**Reply by BenWestgate** [reply]
>So maybe this is too footgunny.

I agree and it lacks spend privacy for the old multisig type which has a legitimate use-case to know which keys signed.

> Thanks for the tip about BIP 85 BTW! I have wanted to do this but didn't realize there was a standard way.

You're welcome.

First thing I did when I heard of BIP85 is write a script to find BIP85 child private extended masterkeys with fingerprints whose first 20-bits match a codex32 ID.

Intended to create a "panic" wallet with less funds and boring transactions that could fool an attacker who knows 20-bits of bip32 fingerprint from a codex32 id.

Takes about a minute per xprv found with a 20-bit match using 4 cores. Indexes don't need storage.

> They should instead store their full watch-only descriptor, but gpg-encrypt it to their own key using BIP85.

Sure and in multi-party, threshold signature, at least N - T others need the descriptor gpg-encrypted to their key also or you block recovery paths.

You probably want to encrypt it to every key. Unless there's a good reason to exclude like a blabber mouth heir or a less trustworthy wallet vendor has it. Subject encrypting to at minimum N - T + 1 keys.

"The *first* key you own" won't work if the threshold is 2 or decays and you only have your *2nd* and *3rd* key? The policy would let you spend but you can't if your descriptor won't decrypt.

I suppose that could be deliberate to prevent collusion but seems that should be explicit in the policy, not subject to change if the cleartext descriptor leaks.

> a specific BIP32 path from it.

Covered by BIP85.

**Reply by apoelstra** [reply]
>You probably want to encrypt it to every key

Ah, good thinking. Agreed, and agree with your reasoning.

>Covered by BIP85.

Phew :) I hate needing (and lacking) a standard for tiny details like this.

**Comment by BenWestgate** [comment]
>The descriptor alphabet is different from the bech32 alphabet (descriptors have a 96-char alphabet which is broken down into the 32-element bech32 alphabet internally) which would complicate all the SSS processes.

That'd be cool but splitting the `xpub` is pretty good privacy when the script is common.

For unique scripts, an SSS upgrade to the `importdescriptors` command should probably come some day.

If we do it for seeds, we should do it for extended pubkeys and descriptors. Both are secrets just to different degrees.
